### PR TITLE
[STORM-1376] Performance slowdown due excessive zk connections and log-debugging

### PR DESCRIPTION
**Pull request by redsanket**
Uneccessary creation of client and closing of connections puts load on zookeeper.
Having a single client connection to perform reads helps

### CODE REVIEW

**Comment by revans2** [comment]
@redsanket did you reproduce the original issue?  Do we need any synchronization around zk client now that we don't create a new one each time?

**Comment by redsanket** [comment]
@revans2 I have launched around 5 topologies on 5 supervisor nodes and could see lot of connections being made to zookeeper and event processing slowing down. But I could not observe zookeeper slowing down to a point it could not respond. I will launch more topologies to make sure it really slows down. I realized in this scenario that it is only performing reads and as per zookeeper implementation I presume connections are thread safe and hence one connection to read information would be sufficient I guess. An interesting point came up with synchronization here, I looked at the localizer code and I have seen downloads to the supervisor are performed by locking but updates are not therefore to support nimbus HA checkForUpdate and checkForDownload might have to be synchronized. Will make sure by running it with more topologies. Thanks for the review

**Comment by redsanket** [comment]
@revans2 Would Automatic backpressure make a difference, as it is turned on now by default and do we need to use pace maker along with it as I guess this change improves to remove the number of connections and speed up a little but on the whole it is not doing good when compared to 0.10.x branch when automatic backpressure is on and running with no pace maker

**Comment by redsanket** [comment]
STORM-1215 has done some changes to Async Logging, I removed those commits and tested it again and now 0.10.x and 0.11 current version are doing similar

File: conf/defaults.yaml [hunk 1]
```diff
@@ -179,7 +179,7 @@ task.refresh.poll.secs: 10
 task.credentials.poll.secs: 30
 
 # now should be null by default
-topology.backpressure.enable: true
+topology.backpressure.enable: false
```
Comment (revans2) [inline conf/defaults.yaml]: Do we want this off by default?

File: storm-core/test/clj/backtype/storm/nimbus_test.clj [hunk 2]
```diff
@@ -1238,10 +1238,11 @@
   (testing "nimbus-data uses correct ACLs"
     (let [scheme "digest"
           digest "storm:thisisapoorpassword"
-          auth-conf {STORM-ZOOKEEPER-AUTH-SCHEME scheme
+          auth-conf (merge (read-storm-config)
+                    {STORM-ZOOKEEPER-AUTH-SCHEME scheme
                      STORM-ZOOKEEPER-AUTH-PAYLOAD digest
                      STORM-PRINCIPAL-TO-LOCAL-PLUGIN "backtype.storm.security.auth.DefaultPrincipalToLocal"
-                     NIMBUS-THRIFT-PORT 6666}
+                     NIMBUS-THRIFT-PORT 6666})
```
Comment (ppoulosk) [inline storm-core/test/clj/backtype/storm/nimbus_test.clj]: I know that the code was already there, but is it possible to use an ephemeral port here?

**Comment by revans2** [comment]
Overall the patch looks good, just curious why we turned off backpressure by default.

File: conf/defaults.yaml [hunk 3]
```diff
@@ -179,7 +179,7 @@ task.refresh.poll.secs: 10
...
 backpressure.disruptor.high.watermark: 0.9
 backpressure.disruptor.low.watermark: 0.4
```
Comment (rfarivar) [inline conf/defaults.yaml]: How about adding a comment describing why the choice of 0.4 and 0.9?

File: storm-core/test/clj/backtype/storm/nimbus_test.clj (hunk 2 above)
Comment (rfarivar) [inline storm-core/test/clj/backtype/storm/nimbus_test.clj]: Ditto.

File: conf/defaults.yaml (hunk 1 above)
Comment (redsanket) [inline conf/defaults.yaml]: Oh Im sorry I was testing few scenarios might have come along with my last commit. Will change it back

**Comment by hustfxj** [comment]
yes, others are good, but why we turned off backpressure by default.

File: conf/defaults.yaml (hunk 3 above)
Comment (zhuoliu) [inline conf/defaults.yaml]: This low/high watermark default values are now more of empirical values based on a limited set of topologies. We may need more experiments and theoretical analysis (based on representative queue draining speed, callback response time etc.) for choices of these watermark percentages.

File: storm-core/test/clj/backtype/storm/nimbus_test.clj (hunk 2 above)
Comment (redsanket) [inline storm-core/test/clj/backtype/storm/nimbus_test.clj]: All the test cases in this file have a similar port, will create a JIRA to tackle this issue separately as I feel it is not a part of this change

**Comment by revans2** [comment]
+1

**Comment by ppoulosk** [comment]
@redsanket, Agreed.  thanks for filing the Jira.  +1, NB

**Comment by knusbaum** [comment]
+1

**Pull request merged**